connect/transactionRequest.ts

```diff
@@ -0,0 +1,141 @@
+  return hashMode === AddressHashMode.SerializeP2SH || hashMode === AddressHashMode.SerializeP2WSH ? true : false;
+};
+
+const cleanMemoString = (memo: string): string => memo.replace('\u0000', '');
+
+function encodePostConditions(postConditions: PostCondition[]) {
+  return postConditions.map((pc) => buf2hex(serializePostCondition(pc)));
+}
+
+export const txPayloadToRequest = (
+  stacksTransaction: StacksTransaction,
+  stxAddress?: string,
+  attachment?: string,
+): TransactionPayload => {
+  const { payload, auth, postConditions, postConditionMode, anchorMode } = stacksTransaction;
+  const encodedPostConditions = encodePostConditions(postConditions.values as PostCondition[]);
+  const transactionRequest = {
+    attachment,
+    stxAddress,
+    sponsored: auth.authType === AuthType.Sponsored,
+    nonce: Number(auth.spendingCondition.nonce),
+    fee: Number(auth.spendingCondition.fee),
+    postConditions: encodedPostConditions,
+    postConditionMode: postConditionMode,
+    anchorMode: anchorMode,
+  } as TransactionPayload;
+  switch (payload.payloadType) {
+    case PayloadType.TokenTransfer:
+      const memo = cleanMemoString(payload.memo.content);
+      (transactionRequest as STXTransferPayload).txType = TransactionTypes.STXTransfer;
+      (transactionRequest as STXTransferPayload).recipient = cvToValue(payload.recipient, true);
+      (transactionRequest as STXTransferPayload).amount = new BigNumber(Number(payload.amount))
+        .shiftedBy(-STX_DECIMALS)
+        .toNumber()
+        .toLocaleString('en-US', { maximumFractionDigits: STX_DECIMALS });
+      (transactionRequest as STXTransferPayload).memo = memo;
+      break;
+    case PayloadType.ContractCall:
+      (transactionRequest as ContractCallPayload).txType = TransactionTypes.ContractCall;
+      (transactionRequest as ContractCallPayload).contractName = payload.contractName.content;
+      (transactionRequest as ContractCallPayload).contractAddress = addressToString(payload.contractAddress);
+      (transactionRequest as ContractCallPayload).functionArgs = payload.functionArgs.map((arg) =>
+        Buffer.from(serializeCV(arg)).toString('hex'),
+      );
+      (transactionRequest as ContractCallPayload).functionName = payload.functionName.content;
+      break;
+    case PayloadType.SmartContract:
+    case PayloadType.VersionedSmartContract:
+      (transactionRequest as ContractDeployPayload).txType = TransactionTypes.ContractDeploy;
+      (transactionRequest as ContractDeployPayload).contractName = payload.contractName.content;
+      (transactionRequest as ContractDeployPayload).codeBody = payload.codeBody.content;
+      (transactionRequest as any).clarityVersion = (
+        stacksTransaction.payload as VersionedSmartContractPayload
+      ).clarityVersion;
+      break;
+    default:
+      throw new Error('Unsupported tx type');
+  }
+
+  return transactionRequest;
+};
```

Comment:
this way of using `as` will mask a lot of potential type errors and kind of make the typed return useless.

if you write it more like this, then the consuming function also knows which specific type it is,
and as you write this function, you can still see what missing properties there are from each of your return types

```suggestion
    const transactionRequest: Partial<TransactionPayload> = {
      attachment,
      stxAddress,
      sponsored: auth.authType === AuthType.Sponsored,
      nonce: Number(auth.spendingCondition.nonce), // this is deprecated
      fee: Number(auth.spendingCondition.fee),
      postConditions: encodedPostConditions,
      postConditionMode: postConditionMode,
      anchorMode: anchorMode,
    };
    switch (payload.payloadType) {
      case PayloadType.TokenTransfer:
E       const stxTransferPayload: STXTransferPayload = { // E: Type '{ txType: TransactionTypes.STXTransfer; recipient: any; amount: string; memo: string; publicKey?: string | undefined; functionArgs?: string[] | undefined; sponsored?: boolean | undefined; ... 12 more ...; nonce?: number | undefined; } | { ...; } |
          ...transactionRequest,
          txType: TransactionTypes.STXTransfer,
          recipient: cvToValue(payload.recipient, true),
          amount: new BigNumber(Number(payload.amount))
            .shiftedBy(-STX_DECIMALS)
            .toNumber()
            .toLocaleString('en-US', { maximumFractionDigits: STX_DECIMALS }),
          memo: cleanMemoString(payload.memo.content),
        };
        return stxTransferPayload;
      case PayloadType.ContractCall:
E       const contractCallPayload: ContractCallPayload = { // E: Type '{ txType: TransactionTypes.ContractCall; contractName: string; contractAddress: string; functionArgs: string[]; functionName: string; publicKey?: string | undefined; sponsored?: boolean | undefined; ... 9 more ...; nonce?: number | undefined; }
          ...transactionRequest,
          txType: TransactionTypes.ContractCall,
          contractName: payload.contractName.content,
          contractAddress: addressToString(payload.contractAddress),
          functionArgs: payload.functionArgs.map((arg) =>
            Buffer.from(serializeCV(arg)).toString('hex'),
          ),
          functionName: payload.functionName.content,
        };
        return contractCallPayload;
      case PayloadType.SmartContract:
      case PayloadType.VersionedSmartContract:
W       const contractDeployPayload: ContractDeployPayload = { // W: 'contractDeployPayload' is assigned a value but never used.
          ...transactionRequest,
          txType: TransactionTypes.ContractDeploy,
          contractName: payload.contractName.content,
          codeBody: payload.codeBody.content,
E         clarityVersion: payload.clarityVersion, // E: Type '{ txType: TransactionTypes.ContractDeploy; contractName: string; codeBody: string; clarityVersion: any; publicKey?: string | undefined; functionArgs?: string[] | undefined; sponsored?: boolean | undefined; ... 11 more ...; nonce?: number | undefined; } |
        };
      default:
        throw new Error('Unsupported tx type');
    }
  };
};
```